feat(ci): Support for install via CI (#135)
GoogleCloudPlatform/spinnaker-for-gcp: ci/JENKINS.md

# Using Jenkins to install Spinnaker for GCP

You can use Jekins to install Spinnaker for GCP. The Jenkins agent executing the job must be installed on a Unix-like operating system. 

The following section assumes you have an existing Jenkins server. If not, consider one of the [Jenkins solutions on Google Cloud](https://cloud.google.com/jenkins/).

## Jenkins on GCP

> If your Jenkins server is running on GCP, follow [best practices](https://cloud.google.com/compute/docs/access/create-enable-service-accounts-for-instances#best_practices) for managing its service account. Your Jenkins server must have full access to all Google Cloud APIs to successfully install Spinnaker for GCP. See the [Compute Engine documentation](https://cloud.google.com/compute/docs/access/create-enable-service-accounts-for-instances#changeserviceaccountandscopes) for guidance on how to modify an instance's Google Cloud API access scopes.

Reviewer: :+1:

- Kubernetes Engine Admin - roles/container.admin 
- Create Service Accounts - roles/iam.serviceAccountCreator
- Pub/Sub Editor - roles/pubsub.editor
- Cloud Memorystore Redis Admin - roles/redis.admin
- Service Usage Admin - roles/serviceusage.serviceUsageAdmin
- Source Repository Administrator - roles/source.admin
- Storage Admin - roles/storage.admin
- Project IAM Admin - roles/resourcemanager.projectIamAdmin
- Service Account User - roles/iam.serviceAccountUser

These roles can be enabled through the IAM UI or with [gcloud](https://cloud.google.com/sdk/gcloud/reference/projects/add-iam-policy-binding).

## Enable Cloud Resource Manager API

The Cloud Resource Manager API must be enabled for Jenkins to successfully retrieve IAM policies. Enable it for your project by visiting the below URL and substituting your project number.

https://console.developers.google.com/apis/api/cloudresourcemanager.googleapis.com/overview?project=[MY-PROJECT-NUMBER]

## Properties file

To get Jenkins to install Spinnaker for GCP, you need to generate a properties file and make it available to Jenkins:

1. Run [setup_properties.sh](../scripts/install/setup_properties.sh).
 

> 1. Make the resulting properties file available to the installation script as it executes the Jenkins job. In this example, the properties file has been uploaded to the Jenkins server using the [Credentials plug-in](https://wiki.jenkins.io/display/JENKINS/Credentials+Plugin). It is made accessible to the job by binding the file to the PROPERTIES variable. Alternatively, you can use a secrets-management solution (like [Hashicorp Vault](https://www.vaultproject.io/) and configure Jenkins to read it from there.

Reviewer 2: Missing closing parens after `Vault`.

### Configure the Jenkins job

Once the dependencies are fulfilled, follow these steps to configure a job to install Spinnaker for GCP. 

1. Create a `New Item` from the Jenkins menu and select a `Freestyle project`. 
2. From the configuration screen, enable `Color ANSI Console Output` and set the `ANSI color map` to `xterm`.
3. Add an `Execute Shell` build step.
4. Configure the build step to retrieve and execute the `setup.sh` script.

```shell
#!/usr/bin/env bash

set -e

git clone https://github.com/GoogleCloudPlatform/spinnaker-for-gcp.git
git config --global user.name "jenkins-user"
git config --global user.email "jenkins-user@example.com"

PARENT_DIR=$WORKSPACE PROPERTIES_FILE=$PROPERTIES CI=true $WORKSPACE/spinnaker-for-gcp/scripts/install/setup.sh
```

In the above example, the Git `user.name` and `user.email` must be configured before you run `setup.sh`. Git operations can also be managed using the [Jenkins Git plugin](https://plugins.jenkins.io/git).

`setup.sh` requires several variables to be passed in:

- `PARENT_DIR`: The absolute path for the Jenkins workspace. Jenkins makes this available via `$WORKSPACE`.

> - `PROPERTIES_FILE`: This is the absolute path to the your generated Spinnaker for GCP properties file.

Reviewer 2: s/the your/your

# Installing Spinnaker for GCP on Jenkins

Installing Spinnaker for GCP can be automated with Jenkins. The Jenkins agent executing the job must be installed on a Unix-like operating system. 

The following section assumes the user has an existing Jenkins server. If not, consider one of the [Jenkins solutions on Google Cloud](https://cloud.google.com/jenkins/).

## Jenkins on GCP

> If your Jenkins server is running on GCP, follow [best practices](https://cloud.google.com/compute/docs/access/create-enable-service-accounts-for-instances#best_practices) for managing its service account. Your Jenkins server must have full access to all Google Cloud APIs to successfully install Spinnaker for GCP.

Author: Yeah the instance is given full access to APIs but the actual IAM permissions are restricted by the IAM roles on its service account. That practice is lifted from the [best practices doc](https://cloud.google.com/compute/docs/access/create-enable-service-accounts-for-instances#best_practices). There is another doc that gives the steps for how to give an instance full Cloud API access, do you think it would be helpful to [link to it](https://cloud.google.com/compute/docs/access/create-enable-service-accounts-for-instances#changeserviceaccountandscopes) as well?

> If your Jenkins server is running on GCP, follow [best practices](https://cloud.google.com/compute/docs/access/create-enable-service-accounts-for-instances#best_practices) for managing its service account. Your Jenkins server must have full access to all Google Cloud APIs to successfully install Spinnaker for GCP.

Reviewer: Yes, that sounds like a good idea.

### AnsiColor Plugin

The [AnsiColor Jenkins Plugin](https://plugins.jenkins.io/ansicolor) is required for properly rendering stdout while installing Spinnaker for GCP. Once the plugin has been installed, enable `Color ANSI Console Output` in the build configuration and set the `ANSI color map` to `xterm`.

## Service account

The Jenkins server must be configured with a GCP service account with the following roles:

- Cloud Functions Developer - roles/cloudfunctions.developer
- Compute Network Viewer - roles/compute.networkViewer
- Kubernetes Engine Admin - roles/container.admin 
- Create Service Accounts - roles/iam.serviceAccountCreator
- Pub/Sub Editor - roles/pubsub.editor
- Cloud Memorystore Redis Admin - roles/redis.admin
- Service Usage Admin - roles/serviceusage.serviceUsageAdmin
- Source Repository Administrator - roles/source.admin
- Storage Admin - roles/storage.admin
- Project IAM Admin - roles/resourcemanager.projectIamAdmin
- Service Account User - roles/iam.serviceAccountUser

These roles can be enabled through the IAM UI or with [gcloud](https://cloud.google.com/sdk/gcloud/reference/projects/add-iam-policy-binding).

> ## Properties file

Author: There are several different ways to do this. For the demo, I used the credentials plugin. What if I add some more detail on that:

*In this example, the properties file has been uploaded to the Jenkins server using the [Credentials plugin](https://support.cloudbees.com/hc/en-us/articles/203802500-Injecting-Secrets-into-Jenkins-Build-Jobs). It is made accessible to the job by binding the file to the PROPERTIES variable.*

> ## Properties file

Reviewer: That looks good, Louis. Is it worth mentioning what the other ways are?

"..has been uploaded to the Jenkins... You could also _____..."

> ## Properties file

Author: I edited line 61 to call out the an alternative secrets management solution? Do you think that would be helpful?

> ## Properties file

Reviewer: :+1:

> In order to install Spinnaker for GCP, you must provide the installation script with a properties file. This properties file can be generated by executing [setup_properties.sh](../scripts/install/setup_properties.sh). Once you have a properties file, it must be provided to the installation script during the execution of the Jenkins job.

Reviewer: ```suggestion
To get Jenkins to install Spinnaker for GCP, you need to generate a properties file and make it available to Jenkins:

1. Run [setup_properties.sh](../scripts/install/setup_properties.sh).
 
1. Make the resulting properties file available to the installation script as it executes the Jenkins job.
   You can use the [Credentials plug-in](https://wiki.jenkins.io/display/JENKINS/Credentials+Plugin), or use a secrets-management solution (like [Hashicorp Vault](https://www.vaultproject.io/) and configure Jenkins to read it from there.
```

> ### Jenkins Job Configuration

Reviewer: ```suggestion
### Configure the Jenkins job
```

> Once the aforementioned dependencies have been fulfilled, follow the below steps to configure a job for installing Spinnaker for GCP. 

Reviewer: ```suggestion
Once the aforementioned dependencies have been fulfilled, follow the below steps to configure a job for installing Spinnaker for GCP. 
```
```suggestion
Once the dependencies are fulfilled, follow these steps to configure a job to install Spinnaker for GCP. 
```

1. Create a `New Item` from the Jenkins menu and select a `Freestyle project`. 
2. From the configuration screen, enable `Color ANSI Console Output` and set the `ANSI color map` to `xterm`.
3. Add an `Execute Shell` build step.
4. Configure the build step to retrieve and execute the `setup.sh` script.

```shell
#!/usr/bin/env bash

set -e

git clone https://github.com/GoogleCloudPlatform/spinnaker-for-gcp.git
git config --global user.name "jenkins-user"
git config --global user.email "jenkins-user@example.com"

REPO_PATH=$WORKSPACE PROPERTIES_FILE=$PROPERTIES CI=true $WORKSPACE/spinnaker-for-gcp/scripts/install/setup.sh
```

> In the above example, the git `user.name` and `user.email` are configured prior to running `setup.sh`. This step is required for `setup.sh` to run successfully. Git operations can also be managed using the [Jenkins Git plugin](https://plugins.jenkins.io/git).

Reviewer: ```suggestion
In the above example, the Git `user.name` and `user.email` must be configured before you run `setup.sh`. Git operations can also be managed using the [Jenkins Git plugin](https://plugins.jenkins.io/git).
```

`setup.sh` requires several variables to be passed in:

- `REPO_PATH`: The absolute path for the Jenkins workspace. Jenkins makes this available via `$WORKSPACE`.
- `PROPERTIES_FILE`: This is the absolute path to the your generated Spinnaker for GCP properties file.
- `CI`: This must be set to `true` when running `setup.sh` outside of Cloud Shell.

> Once configured, execute the job to install Spinnaker for GCP. Changes to the properties file can be applied by re-running the job. Additional instructions for how to access or manage the deployed Spinnaker application are available [here](https://cloud.google.com/docs/ci-cd/spinnaker/spinnaker-for-gcp#access_spinnaker).

Reviewer: ```suggestion
1. Execute the job to install Spinnaker for GCP. 

If you change the properties file, apply the change by re-running the job. 

Additional instructions for how to access or manage the deployed Spinnaker application are available [here](https://cloud.google.com/docs/ci-cd/spinnaker/spinnaker-for-gcp#access_spinnaker).
```

Reviewer: If you indent everything between step 4 and this step (Execute the job...), by three spaces, this becomes just another step in the procedure.
